build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoCommand.java

```diff
@@ -0,0 +1,109 @@
+package org.jfrog.build.extractor.go.extractor;
+
+import com.google.common.collect.Lists;
+import org.apache.commons.lang.StringUtils;
+import org.jfrog.build.api.Artifact;
+import org.jfrog.build.api.Build;
+import org.jfrog.build.api.Dependency;
+import org.jfrog.build.api.Module;
+import org.jfrog.build.api.builder.ModuleBuilder;
+import org.jfrog.build.api.util.Log;
+import org.jfrog.build.client.ArtifactoryVersion;
+import org.jfrog.build.extractor.clientConfiguration.ArtifactoryBuildInfoClientBuilder;
+import org.jfrog.build.extractor.clientConfiguration.ArtifactoryClientBuilderBase;
+import org.jfrog.build.extractor.clientConfiguration.client.ArtifactoryBaseClient;
+import org.jfrog.build.util.VersionCompatibilityType;
+import org.jfrog.build.util.VersionException;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+/**
+ * Base class for go build and go publish commands.
+ */
+abstract class GoCommand implements Serializable {
+
+    protected static final String SHA1 = "SHA1";
+    protected static final String MD5 = "MD5";
+    protected static final String LOCAL_GO_MOD_FILENAME = "go.mod";
+    private static final long serialVersionUID = 1L;
+    private static final ArtifactoryVersion MIN_SUPPORTED_ARTIFACTORY_VERSION = new ArtifactoryVersion("6.10.0");
+
+    ArtifactoryClientBuilderBase clientBuilder;
+    ArtifactoryBaseClient client;
```

Comment:
We shouldn't probably manage in this class both ArtifactoryClientBuilderBase and ArtifactoryBaseClient instances. This makes the decision of which one to use harder. We should simplify the implementation by maintaining only one.
In addition, do we really need to store the base class ArtifactoryBaseClient and not ArtifactoryBuildInfoClient? This can save us a few castings in the code (castings should generally be avoided when possible).method 

Also notice that the **preparePrerequisites(deploymentRepo, client)** receives the client as an argument, although it is accessible to the method ch as a class member. We should try to avoid this. Looks to me like the best approach would be not to save the client as a class member and just build it using the builder when needed.